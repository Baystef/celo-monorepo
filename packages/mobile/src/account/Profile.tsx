import { SettingsItemInput } from '@celo/react-components/components/SettingsItem'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import { StackScreenProps } from '@react-navigation/stack'
import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { ScrollView, StyleSheet, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { useDispatch, useSelector } from 'react-redux'
import { setName, setPicture } from 'src/account/actions'
import { nameSelector, pictureSelector } from 'src/account/selectors'
import { showMessage } from 'src/alert/actions'
import CancelButton from 'src/components/CancelButton'
import i18n, { Namespaces } from 'src/i18n'
import { emptyHeader } from 'src/navigator/Headers'
import { Screens } from 'src/navigator/Screens'
import { TopBarTextButton } from 'src/navigator/TopBarButton'
import { StackParamList } from 'src/navigator/types'
import PictureInput from 'src/onboarding/registration/PictureInput'

type Props = StackScreenProps<StackParamList, Screens.Profile>

function Profile({ navigation, route }: Props) {
  const { t } = useTranslation(Namespaces.accountScreen10)
  const [newName, setNewName] = useState(useSelector(nameSelector) ?? '')
  const [newPicture, setNewPicture] = useState(useSelector(pictureSelector) ?? '')

  const dispatch = useDispatch()

  useEffect(() => {
    if (route.params?.save) {
      navigation.setParams({ save: false })
      dispatch(setPicture(newPicture))
      dispatch(setName(newName))
      // TODO: Save these things on CIP8.
      dispatch(showMessage(t('namePictureSaved')))
      navigation.goBack()
    }
  }, [route.params?.save])

  const onPictureChosen = (chosenPicture: string) => {
    setNewPicture(chosenPicture)
  }

  const updateName = (updatedName: string) => {
    setNewName(updatedName)
  }

  return (
    <ScrollView style={styles.container}>
      <SafeAreaView edges={['bottom']}>
        <View style={styles.accountProfile}>
          <PictureInput
            picture={newPicture}
            onPhotoChosen={onPictureChosen}
            backgroundColor={colors.gray3}
          />
        </View>
        <SettingsItemInput
          value={newName ?? t('global:unknown')}
          testID="ProfileEditName"
          title={t('name')}
          placeholder={t('yourName')}
          onValueChange={updateName}
        />
      </SafeAreaView>
    </ScrollView>
  )
}

Profile.navigationOptions = ({ navigation, route }: Props) => {
  const onCancel = () => {
    navigation.goBack()
  }
  const onSave = () => {
    navigation.setParams({ save: true })
  }
  return {
    ...emptyHeader,
    headerTitle: i18n.t('accountScreen10:editProfile'),
    headerLeft: () => <CancelButton onCancel={onCancel} />,
    headerRight: () => (
      <TopBarTextButton title={i18n.t('global:save')} testID="SaveButton" onPress={onSave} />
    ),
  }
}

export default Profile

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  accountProfile: {
    paddingLeft: 10,
    paddingTop: 30,
    paddingRight: 15,
    paddingBottom: 15,
    flexDirection: 'column',
    alignItems: 'center',
  },
  title: {
    ...fontStyles.h2,
    margin: 16,
  },
})
