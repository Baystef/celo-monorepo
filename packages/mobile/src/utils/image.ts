import * as RNFS from 'react-native-fs'
import Logger from 'src/utils/Logger'

const mimeTypeToExtension: { [key: string]: string } = {
  'image/png': 'png',
  'image/x-png': 'png',
  'image/jpeg': 'jpeg',
  'image/pjpeg': 'jpeg',
  'image/jpg': 'jpg', // image/jpg is technically not a mime type, but handling it just in case.
}

// Data URL format: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs
export const saveImageDataUrlToFile = (
  dataUrl: string,
  fileNameWithoutExtension: string
): string => {
  const mimeType = dataUrl
    .split(':')[1]
    .split(',')[0]
    .split(';')[0]
  const extension = mimeTypeToExtension[mimeType] || '.jpg'
  const fileName = `${fileNameWithoutExtension}.${extension}`
  const data = dataUrl
    .split(',')
    .slice(1) // Not sure if there could be more commas in the data, but if there are just join them again.
    .join(',')
  RNFS.writeFile(fileName, data, 'base64')
    .then(() => Logger.info('Image saved successfully'))
    .catch((e) => Logger.error('Error saving image', e))
  return fileName
}
