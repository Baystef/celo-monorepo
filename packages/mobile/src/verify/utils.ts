import { parsePhoneNumber } from '@celo/utils/lib/phoneNumbers'

export function getPhoneNumberState(
  phoneNumber: string,
  countryCallingCode: string,
  countryCodeAlpha2: string
) {
  const phoneDetails = parsePhoneNumber(phoneNumber, countryCallingCode)

  if (phoneDetails) {
    return {
      nationalPhoneNumber: phoneDetails.displayNumber,
      e164Number: phoneDetails.e164Number,
      isValidNumber: true,
      countryCodeAlpha2: phoneDetails.regionCode!,
    }
  } else {
    return {
      nationalPhoneNumber: phoneNumber,
      e164Number: '',
      isValidNumber: false,
      countryCodeAlpha2,
    }
  }
}
