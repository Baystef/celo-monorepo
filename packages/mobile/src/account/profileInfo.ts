import { ensureLeading0x, eqAddress, Err, normalizeAddressWith0x, Ok, Result } from '@celo/base'
import { Address, ContractKit } from '@celo/contractkit'
import {
  FetchError,
  InvalidSignature,
  OffchainDataWrapper,
  OffchainErrors,
} from '@celo/identity/lib/offchain-data-wrapper'
import { PrivateNameAccessor } from '@celo/identity/lib/offchain/accessors/name'
import { PrivatePictureAccessor } from '@celo/identity/lib/offchain/accessors/pictures'
import { buildEIP712TypedData, resolvePath } from '@celo/identity/lib/offchain/utils'
import { privateKeyToAddress, toChecksumAddress } from '@celo/utils/lib/address'
import { recoverEIP712TypedDataSigner } from '@celo/utils/lib/signatureUtils'
import { UnlockableWallet } from '@celo/wallet-base'
import { SignedPostPolicyV4Output } from '@google-cloud/storage'
// Use targetted import otherwise the RN FormData gets used which doesn't support Buffer related functionality
import FormData from 'form-data/lib/form_data'
import * as t from 'io-ts'
import RNFS from 'react-native-fs'
import { call, put, select } from 'redux-saga/effects'
import { profileUploaded } from 'src/account/actions'
import { isProfileUploadedSelector, nameSelector, pictureSelector } from 'src/account/selectors'
import { extensionToMimeType, getDataURL, saveRecipientPicture } from 'src/utils/image'
import Logger from 'src/utils/Logger'
import { getContractKit, getWallet } from 'src/web3/contracts'
import { currentAccountSelector, dataEncryptionKeySelector } from 'src/web3/selectors'
const TAG = 'account/profileInfo'

const authorizerUrl = 'https://kel03d4ef0.execute-api.eu-west-1.amazonaws.com/dev/authorize'
const valoraMetadataUrl = 'https://d2uaxwjh0f8rcm.cloudfront.net'

async function makeCall(data: any, signature: string): Promise<SignedPostPolicyV4Output[]> {
  const response = await fetch(authorizerUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Signature: signature,
    },
    body: JSON.stringify(data),
  })

  if (response.status >= 400) {
    throw new Error(await response.text())
  }

  return response.json()
}

// Workaround fetch response.arrayBuffer() not working in RN environment
// See https://github.com/facebook/react-native/blob/f96478778cc00da8c11da17f9591dbdf928e7437/Libraries/Blob/FileReader.js#L85
async function responseBuffer(response: Response) {
  const blob = await response.blob()
  return blobToBuffer(blob)
}

// Hacky way to get Buffer from Blob
// Note: this is gonna transfer the whole data over the RN bridge (as base64 encoded string)
// and should be avoided for large files!
function blobToBuffer(blob: Blob) {
  const reader = new FileReader()
  reader.readAsDataURL(blob)
  return new Promise<Buffer>((resolve, reject) => {
    reader.onerror = () => {
      reject(reader.error)
    }
    reader.onloadend = () => {
      const result = reader.result
      if (typeof result !== 'string') {
        reject(new Error('Unexpected result type'))
      } else {
        // Result looks like "data:application/octet-stream;base64,BLMHTkM..."
        // Extract the base64 part
        const base64 = result.substr(result.lastIndexOf(',') + 1)
        resolve(Buffer.from(base64, 'base64'))
      }
    }
  })
}

class UploadServiceDataWrapper implements OffchainDataWrapper {
  signer: Address
  self: Address

  constructor(readonly kit: ContractKit, address: Address) {
    this.signer = this.self = address
  }

  async writeDataTo(
    data: Buffer,
    signature: Buffer,
    dataPath: string
  ): Promise<OffchainErrors | void> {
    const dataPayloads = [data, signature]
    const signedUrlsPayload = [
      {
        path: dataPath,
      },
      {
        path: `${dataPath}.signature`,
      },
    ]

    const hexPayload = ensureLeading0x(
      Buffer.from(JSON.stringify(signedUrlsPayload)).toString('hex')
    )
    const authorization = await this.kit.getWallet().signPersonalMessage(this.signer, hexPayload)
    const signedUrls = await makeCall(signedUrlsPayload, authorization)
    await Promise.all(
      signedUrls.map(({ url, fields }, i) => {
        const formData = new FormData()
        for (const name of Object.keys(fields)) {
          formData.append(name, fields[name])
        }
        formData.append('file', dataPayloads[i])

        return fetch(url, {
          method: 'POST',
          headers: formData.getHeaders(),
          // Use getBuffer() which ends up transferring the body as base64 data over the RN bridge
          // because RN doesn't support Buffer inside FormData
          // See https://github.com/facebook/react-native/blob/b26a9549ce2dffd1d0073ae13502830459051c27/Libraries/Network/convertRequestBody.js#L34
          // and https://github.com/facebook/react-native/blob/b26a9549ce2dffd1d0073ae13502830459051c27/Libraries/Network/FormData.js
          body: formData.getBuffer(),
        }).then((x) => {
          if (!x.ok) {
            Logger.error(TAG + '@writeDataTo', 'Error uploading ' + x.headers.get('location'))
          }
          return x.text()
        })
      })
    )
  }

  async readDataFromAsResult<DataType>(
    account: Address,
    dataPath: string,
    _checkOffchainSigners: boolean,
    type?: t.Type<DataType>
  ): Promise<Result<Buffer, OffchainErrors>> {
    let dataResponse, signatureResponse

    const accountRoot = `${valoraMetadataUrl}/${toChecksumAddress(account)}`
    try {
      ;[dataResponse, signatureResponse] = await Promise.all([
        fetch(resolvePath(accountRoot, dataPath)),
        fetch(resolvePath(accountRoot, `${dataPath}.signature`)),
      ])
    } catch (error) {
      return Err(new FetchError(error))
    }

    if (!dataResponse.ok) {
      return Err(new FetchError(new Error(dataResponse.statusText)))
    }
    if (!signatureResponse.ok) {
      return Err(new FetchError(new Error(signatureResponse.statusText)))
    }

    const [dataBody, signatureBody] = await Promise.all([
      responseBuffer(dataResponse),
      responseBuffer(signatureResponse),
    ])

    const body = Buffer.from(dataBody)
    const signature = ensureLeading0x(Buffer.from(signatureBody).toString('hex'))

    const toParse = type ? JSON.parse(body.toString()) : body
    const typedData = await buildEIP712TypedData(this, dataPath, toParse, type)
    const guessedSigner = recoverEIP712TypedDataSigner(typedData, signature)
    if (eqAddress(guessedSigner, account)) {
      return Ok(body)
    }

    return Err(new InvalidSignature())
  }
}

// requires that the DEK has already been registered
export function* uploadProfileInfo() {
  const isAlreadyUploaded = yield select(isProfileUploadedSelector)
  if (isAlreadyUploaded) {
    return
  }
  try {
    yield call(unlockDEK, true)
    // yield call(addMetadataClaim)
    yield call(uploadNameAndPicture)

    yield put(profileUploaded())
  } catch (e) {
    Logger.error(TAG + '@uploadProfileInfo', 'Error uploading profile', e)
  }
}

// TODO: make metadata claim when registering account info, and use fetched metadata url when reading data

// export function* addMetadataClaim() {
//   try {
//     const contractKit = yield call(getContractKit)
//     const account = yield select(currentAccountSelector)
//     const metadata = IdentityMetadataWrapper.fromEmpty(account)
//     yield call(
//       [metadata, 'addClaim'],
//       createStorageClaim(BUCKET_URL),
//       NativeSigner(contractKit.web3.eth.sign, account)
//     )
//     Logger.info(TAG + '@addMetadataClaim' + 'created storage claim on chain')
//     yield call(writeToGCPBucket, metadata.toString(), `${account}/metadata.json`)
//     Logger.info(TAG + '@addMetadataClaim' + 'uploaded metadata.json')
//     const accountsWrapper: AccountsWrapper = yield call([
//       contractKit.contracts,
//       contractKit.contracts.getAccounts,
//     ])
//     const setAccountTx = accountsWrapper.setMetadataURL(`${BUCKET_URL}${account}/metadata.json`)
//     const context = newTransactionContext(TAG, 'Set metadata URL')
//     yield call(sendTransaction, setAccountTx.txo, account, context)
//     Logger.info(TAG + '@addMetadataClaim' + 'set metadata URL')
//   } catch (error) {
//     Logger.error(`${TAG}/addMetadataClaim`, 'Could not add metadata claim', error)
//     throw error
//   }
// }

export function* uploadNameAndPicture() {
  const contractKit = yield call(getContractKit)
  const account = yield select(currentAccountSelector)
  const offchainWrapper = new UploadServiceDataWrapper(contractKit, account)

  const name = yield select(nameSelector)
  const nameAccessor = new PrivateNameAccessor(offchainWrapper)
  let writeError = yield call([nameAccessor, 'write'], { name }, [])
  Logger.info(TAG + 'uploadNameAndPicture', 'uploaded profile name')

  const pictureUri = yield select(pictureSelector)
  if (pictureUri) {
    const data = yield call(RNFS.readFile, pictureUri, 'base64')
    const mimeType = extensionToMimeType[pictureUri.split('.').slice(-1)] || 'image/jpeg'
    const dataURL = getDataURL(mimeType, data)
    const pictureAccessor = new PrivatePictureAccessor(offchainWrapper)
    writeError = yield call([pictureAccessor, 'write'], dataURL, [])
    Logger.info(TAG + 'uploadNameAndPicture', 'uploaded profile picture')
  }

  if (writeError) {
    Logger.error(TAG + '@uploadNameAndPicture', writeError)
    throw Error('Unable to write data')
  }
}

// this function gives permission to the recipient to view the user's profile info
export function* uploadSymmetricKeys(recipientAddresses: string[]) {
  // TODO: check if key for user already exists, skip if yes
  const account = yield select(currentAccountSelector)
  const contractKit = yield call(getContractKit)
  yield call(unlockDEK)
  const offchainWrapper = new UploadServiceDataWrapper(contractKit, account)

  const nameAccessor = new PrivateNameAccessor(offchainWrapper)
  let writeError = yield call([nameAccessor, 'writeKeys'], recipientAddresses)

  const pictureUri = yield select(pictureSelector)
  if (pictureUri) {
    const pictureAccessor = new PrivatePictureAccessor(offchainWrapper)
    writeError = yield call([pictureAccessor, 'writeKeys'], recipientAddresses)
  }

  Logger.info(TAG + '@uploadSymmetricKeys', 'uploaded symmetric keys for ' + recipientAddresses)

  if (writeError) {
    Logger.error(TAG + '@uploadSymmetricKeys', writeError)
    throw Error('Unable to write keys')
  }
}

export function* getProfileInfo(address: string) {
  // TODO: check if we already have profile info of address
  const account = yield select(currentAccountSelector)
  const contractKit = yield call(getContractKit)
  yield call(unlockDEK)
  const offchainWrapper = new UploadServiceDataWrapper(contractKit, account)

  const nameAccessor = new PrivateNameAccessor(offchainWrapper)
  try {
    const name = yield call([nameAccessor, 'read'], address)

    const pictureAccessor = new PrivatePictureAccessor(offchainWrapper)
    let picturePath
    try {
      const pictureObject = yield call([pictureAccessor, 'read'], address)
      const pictureURL = pictureObject.toString()
      picturePath = yield call(saveRecipientPicture, pictureURL, address)
    } catch (error) {
      Logger.warn("can't fetch picture for", address)
    }
    return { name: name.name, thumbnailPath: picturePath }
  } catch (error) {
    Logger.warn("can't fetch name for", address)
  }
}

function* unlockDEK(addAccount = false) {
  const privateDataKey: string | null = yield select(dataEncryptionKeySelector)
  if (!privateDataKey) {
    throw new Error('No data key in store. Should never happen.')
  }
  const dataKeyaddress = normalizeAddressWith0x(
    privateKeyToAddress(ensureLeading0x(privateDataKey))
  )
  const wallet: UnlockableWallet = yield call(getWallet)
  if (addAccount) {
    try {
      yield call([wallet, wallet.addAccount], privateDataKey, '')
    } catch (e) {
      Logger.warn('Tried adding DEK to geth wallet when it already exists')
    }
  }
  yield call([wallet, wallet.unlockAccount], dataKeyaddress, '', 0)
}
