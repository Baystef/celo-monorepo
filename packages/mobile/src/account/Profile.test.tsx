import * as React from 'react'
import 'react-native'
import { fireEvent, render } from 'react-native-testing-library'
import { Provider } from 'react-redux'
import { saveNameAndPicture } from 'src/account/actions'
import Profile from 'src/account/Profile'
import { Screens } from 'src/navigator/Screens'
import { RootState } from 'src/redux/reducers'
import { createMockStore, getMockStackScreenProps } from 'test/utils'

describe('Profile', () => {
  const store = createMockStore({})
  it('renders correctly', () => {
    const { toJSON } = render(
      <Provider store={store}>
        <Profile {...getMockStackScreenProps(Screens.Profile, { save: false })} />
      </Provider>
    )
    expect(toJSON()).toMatchSnapshot()
  })

  describe('when updating name', () => {
    const name = 'New Name'
    it('edits name', () => {
      const { getByDisplayValue, update } = render(
        <Provider store={store}>
          <Profile {...getMockStackScreenProps(Screens.Profile, { save: false })} />
        </Provider>
      )
      const input = getByDisplayValue((store.getState() as RootState).account.name!)
      fireEvent.changeText(input, name)
      expect(store.getActions().length).toEqual(0)

      update(
        <Provider store={store}>
          <Profile {...getMockStackScreenProps(Screens.Profile, { save: true })} />
        </Provider>
      )
      expect(store.getActions()).toEqual(expect.arrayContaining([saveNameAndPicture(name, null)]))
    })
  })
})
