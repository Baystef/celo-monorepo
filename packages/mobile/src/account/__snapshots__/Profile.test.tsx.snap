// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile renders correctly 1`] = `
<RCTScrollView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <RNCSafeAreaView
      edges={
        Array [
          "bottom",
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "paddingBottom": 15,
            "paddingLeft": 10,
            "paddingRight": 15,
            "paddingTop": 30,
          }
        }
      >
        <RNGestureHandlerButton
          collapsable={false}
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "#B4B9BD",
                "borderRadius": 60,
                "height": 120,
                "justifyContent": "center",
                "opacity": 1,
                "width": 120,
              }
            }
          >
            <Text
              style={Object {}}
            >
              addPhoto
            </Text>
          </View>
        </RNGestureHandlerButton>
        <Modal
          animationType="none"
          hardwareAccelerated={false}
          hideModalContentWhileAnimating={false}
          onModalHide={[Function]}
          onModalWillHide={[Function]}
          onModalWillShow={[Function]}
          onRequestClose={[Function]}
          scrollHorizontal={false}
          scrollOffset={0}
          scrollOffsetMax={0}
          scrollTo={null}
          statusBarTranslucent={true}
          supportedOrientations={
            Array [
              "portrait",
              "landscape",
            ]
          }
          swipeThreshold={100}
          transparent={true}
          visible={false}
        >
          <View
            accessible={true}
            focusable={true}
            forwardedRef={[Function]}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "black",
                "bottom": 0,
                "height": 1334,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 750,
              }
            }
          />
          <View
            forwardedRef={[Function]}
            hideModalContentWhileAnimating={false}
            onModalHide={[Function]}
            onModalWillHide={[Function]}
            onModalWillShow={[Function]}
            pointerEvents="box-none"
            scrollHorizontal={false}
            scrollOffset={0}
            scrollOffsetMax={0}
            scrollTo={null}
            statusBarTranslucent={true}
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
                "margin": 37.5,
                "transform": Array [
                  Object {
                    "translateY": 0,
                  },
                ],
              }
            }
            supportedOrientations={
              Array [
                "portrait",
                "landscape",
              ]
            }
            swipeThreshold={100}
          >
            <RNCSafeAreaView>
              <View
                style={
                  Array [
                    Object {
                      "backgroundColor": "#FFFFFF",
                      "padding": 16,
                    },
                    Object {
                      "borderRadius": 8,
                    },
                    Object {
                      "elevation": 12,
                      "shadowColor": "rgba(156, 164, 169, 0.4)",
                      "shadowOffset": Object {
                        "height": 2,
                        "width": 0,
                      },
                      "shadowOpacity": 1,
                      "shadowRadius": 12,
                    },
                    Array [
                      Object {
                        "backgroundColor": "#FFFFFF",
                        "maxHeight": "100%",
                        "padding": 24,
                      },
                      Object {
                        "alignItems": "center",
                        "justifyContent": "center",
                        "padding": 0,
                      },
                    ],
                  ]
                }
              >
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "opacity": 1,
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "marginVertical": 16,
                        },
                        Object {
                          "color": "#1AB775",
                          "fontFamily": "Inter-Regular",
                          "fontSize": 18,
                          "lineHeight": 24,
                        },
                      ]
                    }
                  >
                    chooseFromLibrary
                  </Text>
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#EDEEEF",
                        "height": 1,
                        "width": "100%",
                      },
                      Object {},
                    ]
                  }
                />
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "opacity": 1,
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "marginVertical": 16,
                        },
                        Object {
                          "color": "#1AB775",
                          "fontFamily": "Inter-Regular",
                          "fontSize": 18,
                          "lineHeight": 24,
                        },
                      ]
                    }
                  >
                    takePhoto
                  </Text>
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#EDEEEF",
                        "height": 1,
                        "width": "100%",
                      },
                      Object {
                        "backgroundColor": "#B4B9BD",
                      },
                    ]
                  }
                />
                <View
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "opacity": 1,
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "marginVertical": 16,
                        },
                        Object {
                          "color": "#1AB775",
                          "fontFamily": "Inter-SemiBold",
                          "fontSize": 18,
                          "lineHeight": 24,
                        },
                      ]
                    }
                  >
                    global:cancel
                  </Text>
                </View>
              </View>
            </RNCSafeAreaView>
          </View>
        </Modal>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
          }
        }
      >
        <View
          style={
            Object {
              "borderBottomColor": "#EDEEEF",
              "borderBottomWidth": 1,
              "marginLeft": 16,
              "paddingVertical": 16,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingRight": 16,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "justifyContent": "center",
                  },
                ]
              }
            >
              <Text
                style={
                  Object {
                    "color": "#2E3338",
                    "fontFamily": "Inter-Regular",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                name
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  Object {
                    "flex": 0,
                    "justifyContent": "flex-end",
                    "paddingLeft": 16,
                  },
                ]
              }
            >
              <TextInput
                allowFontScaling={true}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="yourName"
                rejectResponderTermination={true}
                style={
                  Array [
                    Object {
                      "color": "#2E3338",
                      "flex": 1,
                      "fontFamily": "Inter-Regular",
                      "fontSize": 16,
                      "lineHeight": 22,
                      "paddingHorizontal": 0,
                      "paddingVertical": 12,
                    },
                    undefined,
                    Array [
                      Object {
                        "color": "#9CA4A9",
                        "flex": 0,
                        "minWidth": 160,
                        "paddingVertical": 0,
                        "textAlign": "right",
                      },
                      Object {
                        "color": "#9CA4A9",
                      },
                    ],
                  ]
                }
                testID="ProfileEditName"
                underlineColorAndroid="transparent"
                value="John Doe"
              />
            </View>
          </View>
        </View>
      </View>
    </RNCSafeAreaView>
  </View>
</RCTScrollView>
`;
